File: src/main/kotlin/math/Rectangle.java
package math;

public class Rectangle {
    
    private final float x1, y1, x2, y2;
    
    public Rectangle(float x1, float y1, float x2, float y2) {
        if (x2 < x1) throw new IllegalArgumentException("x2 must be higher or equal to x1");
        if (y2 < y1) throw new IllegalArgumentException("y2 must be higher or equal to y1");
        this.x1 = x1;
        this.y1 = y1;
        this.x2 = x2;
        this.y2 = y2;
    }

    public Rectangle(float width, float height) {
        if (width < 0.0f) throw new IllegalArgumentException("The width must be a positive number.");
        if (height < 0.0f) throw new IllegalArgumentException("The height must be a positive number.");
        this.x1 = 0.0f;
        this.y1 = 0.0f;
        this.x2 = width;
        this.y2 = height;
    }

    public Rectangle(Vec2 position, float width, float height) {
        if (width < 0.0f) throw new IllegalArgumentException("The width must be a positive number.");
        if (height < 0.0f) throw new IllegalArgumentException("The height must be a positive number.");
        this.x1 = position.getX();
        this.y1 = position.getY();
        this.x2 = position.getX() + width;
        this.y2 = position.getY() + height;
    }

    public Rectangle(Vec2 position, Vec2 size) {
        if (size.getX() < 0.0f) throw new IllegalArgumentException("The x size (width) must be a positive number.");
        if (size.getY() < 0.0f) throw new IllegalArgumentException("The y size (height) must be a positive number.");
        this.x1 = position.getX();
        this.y1 = position.getY();
        this.x2 = position.getX() + size.getX();
        this.y2 = position.getY() + size.getY();
    }

    public float getX1() {
        return x1;
    }

    public float getY1() {
        return y1;
    }

    public float getX2() {
        return x2;
    }

    public float getY2() {
        return y2;
    }
}

File: src/main/kotlin/nodeLogic/nodeLogic2d/Rectangle.kt
package nodeLogic.nodeLogic2d

import math.Vec2

data class Rectangle(val x1: Float, val y1: Float, val x2: Float, val y2: Float) {
    init {
        require(x1 >= 0.0f) { "x1 must be positive." }
        require(y1 >= 0.0f) { "y1 must be positive." }
        require(x2 >= 0.0f) { "x2 must be positive." }
        require(y2 >= 0.0f) { "y2 must be positive." }
    }

    constructor(width: Float, height: Float): this(0.0f, 0.0f, width, height) {
        require(width >= 0.0f) { "The width must be a positive number." }
        require(height >= 0.0f) { "The height must be a positive number" }
    }

    constructor(position: Vec2, width: Float, height: Float): this(position.x, position.y, width + position.x, height + position.y) {
        require(width >= 0.0f) { "The width must be a positive number." }
        require(height >= 0.0f) { "The height must be a positive number" }
    }
}
